walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,113 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
+   * Method that finds the available meeting times given events for the days, and request
+   * for the new meeting.
+   *
+   * @param events: Collection of events already scheduled in the day
+   * @param request: Details of the new meeting which we try to find time for
+   * @return list of potential TimeRanges to schedule the new event
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Return empty list if duration exceeds length of day.
+    if (request.getDuration() >= TimeRange.END_OF_DAY) {
+      return new ArrayList<>();
+    }
+
+    // Unmodifiable lists that we get from the params:
+    List<String> attendees = Collections.unmodifiableList(new ArrayList<String>(request.getAttendees()));
+    List<String> optionalAttendees = Collections.unmodifiableList(new ArrayList<String>(request.getOptionalAttendees()));
+    
+    // Modifiable lists that we populate/sort in this function:
+    ArrayList<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> meetingTimesWithOptionals = new ArrayList<TimeRange>();
+
+    // Lists that are populated; made unmodifiable before returning
+    List<TimeRange> requiredSlots = new ArrayList<TimeRange>();
+    List<TimeRange> withOptionalSlots = new ArrayList<TimeRange>();
+
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+      for (String person : attendees) {
+        if (event.getAttendees().contains(person) && !meetingTimes.contains(eventTime)) {
+          meetingTimes.add(eventTime);
+          meetingTimesWithOptionals.add(eventTime);
+        }
+      }
+
+      for (String person : optionalAttendees) {
+        if (event.getAttendees().contains(person)
+            && !meetingTimesWithOptionals.contains(eventTime)) {
+          meetingTimesWithOptionals.add(eventTime);
+        }
+      }
+    }
+    if (meetingTimes.size() == 0 && meetingTimesWithOptionals.size() == 0) {
```

Comment:
You could use `List#isEmpty` to check if list is empty:
https://docs.oracle.com/javase/8/docs/api/java/util/List.html#isEmpty--